Adopt wider tagging model

winem opened the pull request:
This PR tags built images with the major st2 version and major.minor st2 version and pushs those. 

This was tested with ST2_VERSION variations like 3.3, 3.3.0, 3.4dev, 3.4.0dev

So, although I don't expect all the mentioned cases above being relevant, we're covered for any `ST2_VERSION`s like that. 

One of the challenges with the make limitations was that you can't have two lines like `ifeq ($(RELEASE_VERSION), true)` for the same target. That's why I added dedicated sections at the end of each target handling the new tags.

~I'll keep a last set of `make build` and `make push` tests running while going to bed now and mark this PR as draft until tomorrow. I'll remove the draft flag if these tests are successfull, too.~

And I'll create another PR to add the `latest` tag as soon as this one got merged into the master. 

close #31 
closes https://github.com/StackStorm/st2/issues/5139

winem commented:
Ready for review.

armab reviewed (changes requested):
Thanks. The major/minor version parsing logic looks pretty clean :+1: 


Let's imagine we have the following Docker tags present in Docker hub for the image:
- 3.4 - dynamic, refers to `3.4.1`
- 3.4.0 - immutable
- 3.4.1 - immutable

In this situation, triggering the Docker build + deploy for 3.4.0 shouldn't update the `3` or `3.4` tags.
However triggering the Docker build + deploy for 3.4.1 (or new release) should update those references.
So the script needs to fetch the Docker Hub API, find the latest version present, sort and take that into account.

Otherwise we could push some older/outdated Docker images under dynamic tag (like 3.4) by mistake.


I suggest to look at scripts we had before and adopt it for the new model:
* https://github.com/StackStorm/st2-docker/blob/2c2032595a03394b56fbf9cc16114d823a540471/bin/common.sh
* https://github.com/StackStorm/st2-docker/blob/2c2032595a03394b56fbf9cc16114d823a540471/bin/deploy.sh
* https://github.com/StackStorm/st2-docker/blob/2c2032595a03394b56fbf9cc16114d823a540471/bin/gatekeeper.sh


Let's also include the [VERSIONING.md](https://github.com/StackStorm/st2-docker/blob/2c2032595a03394b56fbf9cc16114d823a540471/VERSIONING.md) in the PR to make it super clear for the users.

winem commented:
Hi, 

thanks for the review & feedback. I probably won't have much time to proceed on this topic this weekend as I have to prepare a talk and a hackathon but I'll proceed latest end of next week. 

My current idea is to have a shell script similar to the ones you linked from the st2-docker repo which takes the version of the current build as argument and compares the provided version with the ones available at Dockerhub. The response of the script will be processed by the Makefile to determine whether the additional tags need to be set or not. 

Let me know if you see any issues with the concept or have better ideas.

armab commented:
@winem Sounds good, I like the logic you described more comparing to what we had before. Thanks for the work in this area!

winem commented:
The script to parse and compare the releases takes a bit more time than expected but it'll also be suitable for the other issue where we want to start using the `latest` tag. 

@armab, can it happen, that we release a new version (i.e. a hotfix) for an older release? Let me describe what I mean: 
Latest release at Dockerhub: 3.4
Latest 3.3-patch release at DH: 3.3.4
New release 3.4.1 comes with new features and let's say an important security fix (-> the image will get the tags `latest`, 3 and 3.4) - no questions so far
Now let's say we port that super important security fix back to 3.3 as well. So we build 3.3.5 with that fix and the image should only have the tags 3.3 and 3.3.5 but neither the `latest` nor `3` tag.

Is this a realistic case or something that we generally avoid and don't do in the st2 project? I hope it's clear what I mean.

Just asking not to miss a test case. :)

armab commented:
Between all these versions:
```
3.3.1
3.3.2
3.3 - refers to 3.3.2
3.4.0
3.4.1
3.4 - refers to 3.4.1
latest - refers to v3.4.1
```
the `latest` should always refer to latest stable version, eg `v3.4.1`, even if we release a security patch for the older `v3.3.3` at some point. And yes, if that happens (security patch for and older version) , - `3.3.3` should update `3.3` tag as well.
I guess that's what you described too.


https://github.com/StackStorm/st2-docker/blob/2c2032595a03394b56fbf9cc16114d823a540471/VERSIONING.md

armab commented:
Look at https://github.com/StackStorm/st2-dockerfiles/tree/v3.2 or https://github.com/StackStorm/st2-dockerfiles/tree/v3.3 branches.
We could rebuild the images for those branches at any time if we add new version or even for existing older version (example https://github.com/StackStorm/st2-dockerfiles/blob/v3.2/Makefile#L1).

It's just a matter of restarting a CircleCI build in that branch (https://app.circleci.com/jobs/github/StackStorm/st2-dockerfiles/628) which will deploy to Docker Hub.

winem commented:
Thanks, all clear now!

winem commented:
All tests passed so far. I'll just add some extra error handling in case the determine_needed_tags.sh fails before removing the `Draft` flag from this PR and call it ready for review.

winem commented on determine_needed_tags.sh:
```diff
@@ -0,0 +1,68 @@
+	exit 1
+fi
+
+# dockerhub lists the tags in ascending order. 1st object = lowest tag; last object = highest tag
```
That's a good point. Until now, I thought that dockerhub will order the tags. I'll run some tests to see what actually happens here. Thanks!

winem replied:
@armab I did a bunch of tests now and made the following observations: 

- the returned tags are always ordered by the tag itself NOT by the upload or notification date 
- the first element of the returned list is always the latest tag 

Do we want to rely on it? There seems to be no efficient way in jq to sort the tags because they are strings and not integers. We could do something like piping the jq output to sort the list by ourselves. Any experience with the Docker API and how stable the interface is? Would you expect changes?

armab replied:
Yes, I think writing the script without order assumptions would put us on a safe side, as Docker Hub API doesn't advertize any order for the returned data set: https://docs.docker.com/registry/spec/api/#tags

Some Docker Registries implement `orderby` query param like https://docs.microsoft.com/en-us/rest/api/containerregistry/manifests/getlist#uri-parameters

winem replied:
Done. We don't rely on the API to provide the tags as an ordered list any longer.

winem commented on determine_needed_tags.sh:
```diff
@@ -0,0 +1,68 @@
+tmp_release_cache_file=/tmp/dockerhub_st2_releases
+
+if [[ ${build_version} =~ ^([0-9]+)\.([0-9]+).?([0-9]*)$ ]]; then
+	build_major=${BASH_REMATCH[1]}
+    build_minor=${BASH_REMATCH[2]}
```
Sorry, saw that just now as well. I'll commit it with spaces only.

winem replied:
Done.

winem commented on determine_needed_tags.sh:
```diff
@@ -0,0 +1,68 @@
+
+component=$1
+build_version=$2
+registry=registry.hub.docker.com
```
Done.

winem commented:
The CI failure is "expected" and I know the cause already. Will fix it in the next days.

winem commented on determine_needed_tags.sh:
```diff
@@ -0,0 +1,68 @@
+component=$1
+build_version=$2
+registry=registry.hub.docker.com
+tmp_release_cache_file=/tmp/dockerhub_st2_releases
```
Currently working on it. I do have something in my sleeve but that needs a bit more of refactoring.

winem replied:
Done. We don't create any files any longer.

winem commented on determine_needed_tags.sh:
```diff
@@ -0,0 +1,68 @@
+		# building a release for a new major version
+		tag_update_flag=2
+	fi
+fi
```
I was able to refactor the least popular part of the script which is the one you refer to. Lines of code were reduced and I hope the logic is more clear now.

Nonetheless I would like to explain the logic to support you during the review and make it easier to maintain the script in the future. Please take a look at the updated version of the script and let me know if you prefer additional comments in the script or a short explanation (maybe incl. the test-cases) in a comment to this PR.

winem commented on determine_needed_tags.sh:
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env bash
+
+component=$1
+build_version=$2
```
Added the comments and a usage info. This has actually lead me to another pitfall. Right now, `make` would fail if we build an image for a new component that is not yet available at dockerhub. 

I'll think about this case a bit more and see how to handle it. 

Please feel free to review the usage info and just suggest/apply changes.

winem commented:
So I addressed all points mentioned during the previous review. 

Remaining ToDos: 
- wait for the reviews to be accepted (or continue the ping pong :) ) 
- handle the case where we build an image for a component that does not have an image pushed to dockerhub yet (this would currently fail)

armab commented on determine_needed_tags.sh:
```diff
@@ -0,0 +1,102 @@
+# 0 = no additional tags to be set
+# 1 = add the major.minor tag
+# 2 = add the tags major and major.minor
+# 3 = add the latest tag (for future use)
```
Let's implement `latest`, as logic is not complete in this script without that.

armab commented on Makefile:
```diff
@@ -30,3 +61,11 @@ push:
+		docker push stackstorm/$$image:${MAJOR}.${MINOR}; \
+		echo -e "\033[32mSuccessfully pushed \033[1mstackstorm/$$image:${MAJOR}.${MINOR}\033[0m\033[32m Docker image for StackStorm version \033[1m${ST2_VERSION}\033[0m"; \
+	done
+endif
```
need new line at the end of file

winem replied:
Already committed and will be pushed with the other additions later today. 

But why is this needed? Is it just about a style convention or linting rules or is there any technical reason for it?

armab replied:
As I'm aware it's just best coding/styling practice, but looks like there are a bunch of technical reasons for that too.

Interesting:
https://stackoverflow.com/questions/5813311/no-newline-at-end-of-file
https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

winem replied:
Thanks, that's helpful! :+1:

armab commented on determine_needed_tags.sh:
```diff
@@ -0,0 +1,102 @@
+  fi
+fi
+
+echo $tag_update_flag
```
need new line at the end of file

winem commented:
The version I just pushed handles the latest tag as well now. I'm quite confident, that it works as expected now. 

I'll do another big round of tests tomorrow evening an create & provide a test protocol of all the test cases. I guess this makes the review easier and spot missing cases - if any.

winem commented:
Now I updated the PR and all comments should be addressed. I also ran a whole bunch of tests and all of them were successfully. In the end I reduced the test plan to a set of a single test case per relevant test case. 

The tests in described in the attached table were made with the following tags being available at dockerhub. I hope this helps. Please let me know if you see another test case and I missed something. 

![Selection_194](https://user-images.githubusercontent.com/1457899/101999172-a98c1880-3cda-11eb-964d-a3f28b30345e.png)

From my point of view, this PR is RFR now and I remove the draft flag.

armab commented on Makefile:
```diff
@@ -1,7 +1,24 @@
+# 0 = no additional tags to be set
+# 1 = add the major.minor tag
+# 2 = add the tags major and major.minor
+# 3 = add the latest tag (for future use)
```
Looks like "for future use" could be removed now as `latest` was impelemented

winem replied:
Done.

armab commented on determine_needed_tags.sh:
```diff
@@ -0,0 +1,146 @@
+  echo "image_version = the st2 release used for your current build (i.e. 3.3.0)"
+  echo 
+  echo "Possible output:"
+  echo "0 = no additional tags to be updated"
+  echo "1 = update the major.minor tag (i.e. 3.3)"
+  echo "2 = update the major and the major.minor tag (i.e. 3 and 3.3)"
+  echo
```
Missing description for the `3` option with the `latest` tag.

winem replied:
Also done. Thanks for the review and feedback!

armab reviewed (commented):
Great work & really serious testing there! :+1: 

Looks good.
Just left a couple of minor comments around docs and would be nice to include the adjusted https://github.com/StackStorm/st2-docker/blob/2c2032595a03394b56fbf9cc16114d823a540471/VERSIONING.md file in the PR for crystal clarity.

As we're really close to `v3.4.0` release, will merge this PR soon.

armab reviewed (changes requested):
I tried the script here.

CircleCI build stage (https://app.circleci.com/pipelines/github/StackStorm/st2-dockerfiles/467/workflows/c60acb93-1427-485d-ae67-3c31200ac3d4/jobs/985) reported the following: 
```
#!/bin/sh -eo pipefail
make build

./determine_needed_tags.sh: line 65: curl: command not found
./determine_needed_tags.sh: line 67: [: -eq: unary operator expected
./determine_needed_tags.sh: line 72: [: -ne: unary operator expected
./determine_needed_tags.sh: line 78: curl: command not found
./determine_needed_tags.sh: line 79: jq: command not found
./determine_needed_tags.sh: line 83: available_releases: bad array subscript
./determine_needed_tags.sh: line 90: [: 3.3.0: unary operator expected
./determine_needed_tags.sh: line 95: [: 3: unary operator expected
./determine_needed_tags.sh: line 65: curl: command not found
./determine_needed_tags.sh: line 67: [: -eq: unary operator expected
./determine_needed_tags.sh: line 72: [: -ne: unary operator expected
./determine_needed_tags.sh: line 78: curl: command not found
./determine_needed_tags.sh: line 79: jq: command not found
./determine_needed_tags.sh: line 83: available_releases: bad array subscript
./determine_needed_tags.sh: line 90: [: 3.3.0: unary operator expected
./determine_needed_tags.sh: line 95: [: 3: unary operator expected
./determine_needed_tags.sh: line 65: curl: command not found
./determine_needed_tags.sh: line 67: [: -eq: unary operator expected
./determine_needed_tags.sh: line 72: [: -ne: unary operator expected
./determine_needed_tags.sh: line 78: curl: command not found
./determine_needed_tags.sh: line 79: jq: command not found
./determine_needed_tags.sh: line 83: available_releases: bad array subscript
./determine_needed_tags.sh: line 90: [: 3.3.0: unary operator expected
./determine_needed_tags.sh: line 95: [: 3: unary operator expected
./determine_needed_tags.sh: line 65: curl: command not found
./determine_needed_tags.sh: line 67: [: -eq: unary operator expected
./determine_needed_tags.sh: line 72: [: -ne: unary operator expected
./determine_needed_tags.sh: line 78: curl: command not found
./determine_needed_tags.sh: line 79: jq: command not found
./determine_needed_tags.sh: line 83: available_releases: bad array subscript
./determine_needed_tags.sh: line 90: [: 3.3.0: unary operator expected
./determine_needed_tags.sh: line 95: [: 3: unary operator expected
./determine_needed_tags.sh: line 65: curl: command not found
./determine_needed_tags.sh: line 67: [: -eq: unary operator expected
./determine_needed_tags.sh: line 72: [: -ne: unary operator expected
./determine_needed_tags.sh: line 78: curl: command not found
./determine_needed_tags.sh: line 79: jq: command not found
./determine_needed_tags.sh: line 83: available_releases: bad array subscript
./determine_needed_tags.sh: line 90: [: 3.3.0: unary operator expected
./determine_needed_tags.sh: line 95: [: 3: unary operator expected
./determine_needed_tags.sh: line 65: curl: command not found
./determine_needed_tags.sh: line 67: [: -eq: unary operator expected
./determine_needed_tags.sh: line 72: [: -ne: unary operator expected
./determine_needed_tags.sh: line 78: curl: command not found
./determine_needed_tags.sh: line 79: jq: command not found
./determine_needed_tags.sh: line 83: available_releases: bad array subscript
./determine_needed_tags.sh: line 90: [: 3.3.0: unary operator expected
./determine_needed_tags.sh: line 95: [: 3: unary operator expected
```

Despite errors script returned `3` which tags the st2 `v3.2.0` images as: `3.2.0`, `3.2`, `3`, `latest`. That's worst case scenario as it might overwrite existing production tags with older/irrelevant versions.

armab commented on determine_needed_tags.sh:
```diff
@@ -0,0 +1,147 @@
+  tag_update_flag=3
+else
+  # building a release for a st2 version that does not match the latest version
+  if [ ${build_major} -le ${latest_major} ]; then
```
Here is what I got when running the script locally: 
```
./determine_needed_tags.sh st2 3.0.0

./determine_needed_tags.sh: line 95: [: latest: integer expression expected
3
```

winem replied:
Ofc, that shouldn't happen. I'll take a look at it tomorrow evening latest and try to provide a fix on the weekend. Will also see how this passed my own tests then.

winem replied:
Oh ok, I see how this happened. It looks like there are images at Dockerhub that do already have the latest tag and this one is now provided as the last item in the list (and no longer the actual latest release i.e. 3.3.0). 

I'll address that shortly.

winem replied:
Ok, this is fixed now. The main issue was the one I already described above: the script expected a release version like 3.3.0 and not `latest`. Now, `latest` as available release is also filtered and no longer returned by jq. 

The more important point is your 2nd one: 
```
./determine_needed_tags.sh: line 95: [: latest: integer expression expected
3
```
This was a glidge in the script due to the "wonderful" (sorry, don't find an appropriate word for it right now :) ) error handling in bash. 

What happend is that the if clause failed due to `latest` not being an integer and falling back to the else clause. I addressed this by changing: 
```
  if [ ${build_major} -le ${latest_major} ]; then # < this one failed
[...]
  else # < so we got here
    # building a release for a new major version
    tag_update_flag=3
  fi
```
to: 
```
  if [ ${build_major} -le ${latest_major} ]; then
[...]
  elif [ ${build_major} -gt ${latest_major} ]; then
    # building a release for a new major version
    tag_update_flag=3
  fi
```
So the worst thing that could happen now if parsing of a returned tag leads to an unwanted result is the error below: 
```
./determine_needed_tags.sh: line 95: [: latest: integer expression expected
./determine_needed_tags.sh: line 141: [: latest: integer expression expected
0
```

So the script returns 0 and no tag would be updated. I'll see if there is a more elegant solution to address this and even avoid (=handle) the 2 error messages.

winem commented:
@armab can we install curl on the CI node or should I switch to another tool like wget? 

For now, I addressed the issues mentioned in my previous comments (already pushed) and improved the error handling regarding the failing curl command (not yet pushed as I'll wait for your feedback regarding the right tool to choose for the HTTP queries).

armab commented:
Sure, feel free to adjust the CircleCI config to install `curl` and `jq`.

winem commented:
Sorry, I forgot that I can do it by myself. Will do it and let you know when it's done. Thanks!

winem commented:
I pushed some improvements on the error handling and additional validation of the variables. The CI run is still expected to fail as I did not push the new circleci config yet. I'll do that once I saw that the errors are handled as expected now.

armab commented:
Thanks!

Interesting that even after the latest changes CircleCI build (https://app.circleci.com/pipelines/github/StackStorm/st2-dockerfiles/474/workflows/5dd3a84c-67f4-43f8-8eb1-5ead13c2cf4c/jobs/992) still reports syntax errors while I guess it should stop at dependencies check:

```
#!/bin/sh -eo pipefail
make build

./determine_needed_tags.sh: line 47: [: curl: binary operator expected
./determine_needed_tags.sh: line 78: curl: command not found
./determine_needed_tags.sh: line 80: [: -eq: unary operator expected
./determine_needed_tags.sh: line 84: [: -ne: unary operator expected
./determine_needed_tags.sh: line 90: curl: command not found
./determine_needed_tags.sh: line 91: jq: command not found
./determine_needed_tags.sh: line 95: available_releases: bad array subscript
./determine_needed_tags.sh: line 106: [: 3.2.0: unary operator expected
./determine_needed_tags.sh: line 111: [: 3: unary operator expected
./determine_needed_tags.sh: line 157: [: 3: unary operator expected
```

winem commented:
Hi, this was caused by the variable expansion of ${missing_packages} in bash and I fixed it. 

But I would love to hear your input regarding a design question. 

Now, the issue is that for curl and jq are missing, so `determine_needed_tags.sh` fails with exit 1 and returns `Requirement(s) not satisfied: curl jq` which becomes the value of the TAG_UPDATE_FLAG in the makefile. Neither the failing script nor it's output is immediately visible in the CI logs. 

The code I pushed just now does not cause the make command to fail but prints an error that no tags were set. Is this ok or do we want the CI run to fail, too? 

I do have a few options in mind to let the make command fail, too, but this requires some refactoring of the makefile. I could get it done this week in the evenings to have it ready before the code freeze for 3.4.

armab commented on Makefile:
```diff
@@ -1,7 +1,24 @@
+# 1 = add the major.minor tag
+# 2 = add the tags major and major.minor
+# 3 = add the tags major, major.minor and latest
+TAG_UPDATE_FLAG := $(shell ./determine_needed_tags.sh st2 ${ST2_VERSION})
```
I think it would be best to just scan the exit code from the script and fail fast if it's non-zero.

winem replied:
Now all failures in the script should be handled and return a message with the "Error:"-Prefix and the makefile checks the output for the Error-Prefix and fails before building any images. 

Another idea was to introduce another return value like -1 in case of an error but then I wouldn't know how to show the error message in the CI logs. If you have an idea to show the output from the script in the CI log please let me know. 

I just pushed a Makefile with 3.3.0 and will change it back to 3.4dev before merging this branch. That's just because the determine tags script is skipped if we build a *dev version and not a release.

armab replied:
Depending on "Error" message as a failure flag adds some knowledge overhead in requirements that could be broken in the future.
What if 3rd Maintainer comes, edit the `determine_needed_tags.sh` script and doesn't add "Error" message for the failure case.

We can at least reinforce it with `determine_needed_tags.sh || echo Error`  to fit your solution.

winem commented:
So, this one failed as expected because curl and jq were not installed: https://app.circleci.com/pipelines/github/StackStorm/st2-dockerfiles/482/workflows/f8a8c4b3-a6f5-493a-a17d-70f88f686d80/jobs/1000

This one for 3.3.0 was successful and tagged the images as expected:
https://app.circleci.com/pipelines/github/StackStorm/st2-dockerfiles/485/workflows/48cac0b0-436b-4d7f-8c83-e34766361f43/jobs/1003

And this one for 3.4dev was successful as well. No additional tags were set because it's a dev-build. So from my side, this should be fine now.

armab commented on Makefile:
```diff
@@ -1,10 +1,35 @@
 # Build all required images (st2 base image plus st2 components)
 .PHONY: build
 build:
+ifeq ($(RELEASE_VERSION), true)
+ifneq ($(shell echo "${TAG_UPDATE_FLAG}" | grep -E "Error:"),)
+	@echo -e "Failed to identify the tags to be set."
+	@echo -e "\033[31mNo images were tagged due to an error when determining the correct tags: ${TAG_UPDATE_FLAG}\033[0m"
+	exit 1
+endif
+endif
```
Looks like the same block repeated in both `build` and `push`.

I think it would be great to create a dedicated makefile target to avoid code duplication.

winem replied:
Good point! Done.

armab reviewed (approved):
Gave it a few more testing cycles and it works really well now.
Thanks for the diligence and extra work catching all the corner cases :+1: 

It was important to make sure we don't accidentally overwrite the wrong Docker tag in prod/release.
